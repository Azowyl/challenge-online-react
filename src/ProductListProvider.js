import React, {useCallback} from 'react';

const ProductListContext = React.createContext({
    productList: [],
    addProduct: () => {},
});

function ProductListProvider({ children }) {
    const [productList, setProductList] = React.useState([]);

    const addProduct = useCallback((product) => {
        setProductList((prevProductList) => {
            return [...prevProductList, product];
        });
    }, [setProductList]);

    return (
        <ProductListContext.Provider value={{ productList: productList, addProduct }}>
            {children}
        </ProductListContext.Provider>
    )
}

export const useProductContext = () => React.useContext(ProductListContext);

export default ProductListProvider;