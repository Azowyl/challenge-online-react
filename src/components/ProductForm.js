import React, {Component, useState} from "react";

import "./components.scss";

function ProductForm() {
  const [productData, setProductData] = useState({
    product: "",
    count: "",
    price: "",
    promotionalPrice: "",
  });

  const handleInputChange = (field, event) => {
    return setProductData({[field]: event.target.value})
  }

  const handleSubmit = (event) => {
    event.preventDefault();
  }

  return (
    <div className="centered">
      <h2>Add new product</h2>
      <div>
        <label className="label" htmlFor="">Product name</label>
        <input name="product" type="text" value={ productData.product } onChange={handleInputChange}/>
        <label className="label" htmlFor="">Items count</label>
        <input name="count" type="text" value={ productData.count } onChange={handleInputChange}/>
        <label className="label" htmlFor="">Price</label>
        <input name="price" type="text" value={ productData.price } onChange={handleInputChange}/>
        <label className="label" htmlFor="">Promotional Price</label>
        <input name="promotionalPrice" type="text" value={ productData.promotionalPrice } onChange={handleInputChange}/>
        <input className="button" type="submit" value="Aceptar" />
      </div>
    </div>
  );
}

export default ProductForm;
