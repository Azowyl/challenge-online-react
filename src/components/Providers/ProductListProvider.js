import React, {useCallback} from 'react';
import usePersistedState from "../../hooks/usePersistedState";

const ProductListContext = React.createContext({
    productList: [],
    addProduct: () => {},
    removeProduct: () => {},
});

function ProductListProvider({ children }) {
    const [productList, setProductList] = usePersistedState("productList", []);

    const addProduct = useCallback((product) => {
        setProductList(prevList => [...prevList, product]);
    }, [setProductList]);

    const removeProduct = useCallback((productName) => {
        setProductList(productList.filter((p) => p.name !== productName));
    }, [setProductList, productList]);

    return (
        <ProductListContext.Provider value={{ productList: productList, addProduct, removeProduct }}>
            {children}
        </ProductListContext.Provider>
    )
}

export const useProductContext = () => React.useContext(ProductListContext);

export default ProductListProvider;