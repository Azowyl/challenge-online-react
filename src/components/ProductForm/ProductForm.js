import React, {useState} from "react";
import {useProductContext} from "../Providers/ProductListProvider";
import Input from "../Input/Input";

import "./productForm.scss";
import BottomSpacer from "../Spacing/BottomSpacer";
import SubmitButton from "../Button/SubmitButton";
import {CopyIcon, MoneyIcon, TagIcon, VolumeIcon} from "@tiendanube/icons";

function ProductForm({location}) {
  const [productData, setProductData] = useState(location.state || {
    name: "",
    count: "",
    price: "",
    promotionalPrice: "",
  });
  const [errors, setErrors] = useState({
    name: "",
    count: "",
    price: "",
    promotionalPrice: "",
  });

  const { addProduct } = useProductContext();

  const handleInputChange = (event) => {
    return setProductData({...productData, [event.target.id]: event.target.value})
  }

  const handleSubmit = (event) => {
    event.preventDefault();

    const errors = getFormErrors();
    if (Object.keys(errors).length > 0) {
        setErrors(errors);
    } else {
        addProduct(productData);
    }
  }

  const getFormErrors = () => {
    let errors = {};
    Object.entries(productData).forEach(([key, value]) => {
      if (!value) {
        errors[key] = "This field is required";
      }
    });

    return errors;
  }

  return (
    <div className="centered">
      <h2>Add new product</h2>
      <form onSubmit={handleSubmit}>
        <Input
            name="name"
            type="text"
            value={productData.name}
            onChange={handleInputChange}
            label="Product name"
            error={errors.name}
            icon={<TagIcon />}
        />
        <BottomSpacer />
        <Input
            name="count"
            type="number"
            value={productData.count}
            onChange={handleInputChange}
            label="Items count"
            error={errors.count}
            icon={<CopyIcon />}
            min={0}
        />
        <BottomSpacer />

        <Input
            name="price"
            type="number"
            value={productData.price}
            onChange={handleInputChange}
            label="Price"
            error={errors.price}
            icon={<MoneyIcon />}
            min={0}
        />
        <BottomSpacer />
        <Input
            name="promotionalPrice"
            type="number"
            value={productData.promotionalPrice}
            onChange={handleInputChange}
            label="Promotional Price"
            error={errors.promotionalPrice}
            icon={<VolumeIcon />}
            min={0}
        />
        <BottomSpacer />

        <SubmitButton label="Aceptar" />
      </form>
    </div>
  );
}

export default ProductForm;
