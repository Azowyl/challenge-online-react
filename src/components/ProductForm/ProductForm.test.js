import React from 'react';
import ProductListProvider from "../Providers/ProductListProvider";
import {render, screen} from "@testing-library/react";
import ProductForm from "./ProductForm";
import userEvent from "@testing-library/user-event";

describe('ProductForm', () => {
    let renderComponent;

    beforeEach(() => {
        renderComponent = () => render(
            <ProductListProvider>
                <ProductForm />
            </ProductListProvider>
        );
    })

    it('renders form correctly', () => {
        expect(renderComponent().asFragment()).toMatchSnapshot();
    });

    describe('error validation', () => {
        beforeEach(async () => {
            renderComponent();

            await userEvent.type(screen.getByLabelText('Product name'), 'Test product');
            await userEvent.type(screen.getByLabelText('Items count'), '5');
            await userEvent.type(screen.getByLabelText('Price'), '21');
            await userEvent.type(screen.getByLabelText('Promotional Price'), '17');
        });

        it.each(['Product name', 'Items count', 'Price', 'Promotional Price'])
        ('when %s is empty, it shows an error message', async (field) => {
            await userEvent.clear(screen.getByLabelText(field));
            await userEvent.click(screen.getByRole('button', {name: 'Aceptar'}));

            expect(screen.getByText('This field is required')).toBeInTheDocument();
        })
    })
});