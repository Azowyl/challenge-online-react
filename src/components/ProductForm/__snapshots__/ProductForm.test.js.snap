// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm renders form correctly 1`] = `
<DocumentFragment>
  <div
    class="centered"
  >
    <h2>
      Add new product
    </h2>
    <form>
      <label
        class="label"
        for="name"
      >
        Product name
      </label>
      <span
        style="flex-direction: row; display: flex; align-items: center;"
      >
        <svg
          aria-hidden="true"
          aria-label="Tag"
          fill="currentColor"
          height="16"
          role="img"
          style="display: inline-block; user-select: none; vertical-align: middle;"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M15 7.2L8.8 1a1.25 1.25 0 00-.88-.34h-6A1.25 1.25 0 00.66 1.91v6A1.25 1.25 0 001 8.8L7.2 15a1.24 1.24 0 00.88.36A1.28 1.28 0 009 15l6-6a1.27 1.27 0 000-1.8zm-6.9 6.89L1.91 7.92v-6h6l6.17 6.17z"
          />
          <ellipse
            cx="4.95"
            cy="4.95"
            rx="1.41"
            ry="1.5"
          />
        </svg>
        <span
          style="width: 100%;"
        >
          <input
            id="name"
            name="name"
            type="text"
            value=""
          />
        </span>
      </span>
      <div
        class="bottom-spacer"
      />
      <label
        class="label"
        for="count"
      >
        Items count
      </label>
      <span
        style="flex-direction: row; display: flex; align-items: center;"
      >
        <svg
          aria-hidden="true"
          aria-label="Copy"
          fill="currentColor"
          height="16"
          role="img"
          style="display: inline-block; user-select: none; vertical-align: middle;"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M13.49 3L10.74.37A1.22 1.22 0 009.86 0h-4a1.25 1.25 0 00-1.22 1.25v11a1.25 1.25 0 001.25 1.25h6.72a1.25 1.25 0 001.25-1.25V3.88a1.22 1.22 0 00-.37-.88zm-.88 9.25H5.89v-11h2.72v2.63a1.25 1.25 0 001.25 1.25h2.75zm0-8.37H9.86V1.25l2.75 2.63z"
          />
          <path
            d="M10.11 14.75H3.39v-11H4V2.5h-.61a1.25 1.25 0 00-1.25 1.25v11A1.25 1.25 0 003.39 16h6.72a1.25 1.25 0 001.25-1.25v-.63h-1.25z"
          />
        </svg>
        <span
          style="width: 100%;"
        >
          <input
            id="count"
            min="0"
            name="count"
            type="number"
            value=""
          />
        </span>
      </span>
      <div
        class="bottom-spacer"
      />
      <label
        class="label"
        for="price"
      >
        Price
      </label>
      <span
        style="flex-direction: row; display: flex; align-items: center;"
      >
        <svg
          aria-hidden="true"
          aria-label="Money"
          fill="currentColor"
          height="16"
          role="img"
          style="display: inline-block; user-select: none; vertical-align: middle;"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M12.32 8a3 3 0 00-2-.7H5.63A1.59 1.59 0 014 5.69a2 2 0 010-.25 1.59 1.59 0 011.63-1.33h4.62a1.59 1.59 0 011.57 1.33h1.5a3.08 3.08 0 00-3.07-2.83H8.67V.31H7.42v2.3H5.63a3.08 3.08 0 00-3.07 2.83 2.09 2.09 0 000 .25 3.07 3.07 0 003.07 3.07h4.74A1.59 1.59 0 0112 10.35a1.86 1.86 0 010 .34 1.59 1.59 0 01-1.55 1.24h-4.7a1.59 1.59 0 01-1.55-1.24H2.69a3.08 3.08 0 003.06 2.73h1.67v2.27h1.25v-2.27h1.7a3.08 3.08 0 003.06-2.73v-.34A3.06 3.06 0 0012.32 8z"
          />
        </svg>
        <span
          style="width: 100%;"
        >
          <input
            id="price"
            min="0"
            name="price"
            type="number"
            value=""
          />
        </span>
      </span>
      <div
        class="bottom-spacer"
      />
      <label
        class="label"
        for="promotionalPrice"
      >
        Promotional Price
      </label>
      <span
        style="flex-direction: row; display: flex; align-items: center;"
      >
        <svg
          aria-hidden="true"
          aria-label="Volume"
          fill="currentColor"
          height="16"
          role="img"
          style="display: inline-block; user-select: none; vertical-align: middle;"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M15.32 7.45h-2v1.06h2V8a4.93 4.93 0 000-.55zM12.91 11l1.56.37a6.28 6.28 0 00.45-1L13.16 10zm2-5.43a6.28 6.28 0 00-.45-1l-1.55.34.25 1zm-4.29-3.06L1.9 6.3v-.69H.65v4.78H1.9v-.71l8.72 3.79v1h1.25V1.5h-1.25zM1.9 8.32v-.65l8.72-3.8v8.24z"
          />
        </svg>
        <span
          style="width: 100%;"
        >
          <input
            id="promotionalPrice"
            min="0"
            name="promotionalPrice"
            type="number"
            value=""
          />
        </span>
      </span>
      <div
        class="bottom-spacer"
      />
      <button
        type="submit"
      >
        Aceptar
      </button>
    </form>
  </div>
</DocumentFragment>
`;
